ENH Add `zero_division` parameter for `accuracy_score`

Jaimin020 opened the pull request:
Introduce the zero_division parameter to the accuracy_score function when y_true and y_pred are empty.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Make zero_division parameter consistent in the different metric #29048 (Task-1)


#### What does this implement/fix? Explain your changes.

I have verified the lengths of the variables "y_true" and "y_pred." If both lengths are 0, I have generated an output based on the value of the "zero_warning" variable.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Jaimin020 committed:
Addition of "zero_division"

Introduce the zero_division parameter to the accuracy_score function when y_true and y_pred are empty.

Jaimin020 committed:
Merge branch 'main' into 29048_task_1

Jaimin020 committed:
Merge branch 'scikit-learn:main' into 29048_task_1

Jaimin020 committed:
Resolve checks

Revise the method for calculating the number of elements in y_pred and y_true.

StefanieSenger commented:
Hi @Jaimin020,

regarding the CodeCov failures: you need to add `accuracy_score` to the tests in `test_classification.py`. The output of the CI seems not very helpful here however. Normally, the precise lines of the uncovered code should show up in the "Files changed" tab.

StefanieSenger commented on sklearn/metrics/_classification.py:
```diff
@@ -154,10 +154,16 @@ def _check_targets(y_true, y_pred):
         "normalize": ["boolean"],
         "sample_weight": ["array-like", None],
+        "zero_division": [
+            Options(Real, {0, 1}),
```
I think you have to include `np.nan` here.

Jaimin020 replied:
Done.

StefanieSenger commented on sklearn/metrics/_classification.py:
```diff
@@ -181,6 +187,12 @@ def accuracy_score(y_true, y_pred, *, normalize=True, sample_weight=None):
+        Sets the value to return when there is a zero division.
+
+        Notes:
+        - If set to "warn", this acts like 0, but a warning is also raised.
```
To my ears "acts like 0" is not very clear. A suggestion:

```suggestion
        - If set to "warn", this behaves like a 0.0 input, but a warning is also 
        raised.
```

Jaimin020 replied:
Done

StefanieSenger commented on sklearn/metrics/_classification.py:
```diff
@@ -218,6 +230,16 @@ def accuracy_score(y_true, y_pred, *, normalize=True, sample_weight=None):
+    len_y_true = _num_samples(y_true)
+    len_y_pred = _num_samples(y_pred)
+
+    if len_y_true == 0 and len_y_pred == 0:
```
Working in my own PR, I was thinking that a try except clause is easier to read for later maintanance, because with defining a condition like `if len_y_true == 0 and len_y_pred == 0` we cannot easily see that this would cause a division by zero lateron.

StefanieSenger reviewed (commented):
I leave you some thoughts, that might be helpful for furthing this PR, @Jaimin020.

The type of this PR would not be Fix, but ENH and you need to add an entry into `doc/whats_new/v1.6.rst` as well.

StefanieSenger commented on sklearn/metrics/_classification.py:
```diff
@@ -218,6 +230,16 @@ def accuracy_score(y_true, y_pred, *, normalize=True, sample_weight=None):
+    if len_y_true == 0 and len_y_pred == 0:
+        score = _check_zero_division(zero_division)
+        if zero_division == "warn":
+            _warn_prf(None, "Predcited", "Accuracy is", 0)
```
```suggestion
            _warn_prf(None, "Predcited", "`accuracy_score` is", 0)
```
I think we should be precise and tell our user where exactly the warning stems from.

Jaimin020 replied:
Done

Jaimin020 committed:
Merge branch 'scikit-learn:main' into 29048_task_1

Jaimin020 committed:
Added test cases and makes a mentioned changes.

Added test cases for the 'zero_division' parameter in the accuracy_score function. Additionally, minor bugs were fixed as noted by the reviewer, and 'doc/whats_new/v1.6.rst' has been updated.

Jaimin020 committed:
Merge branch 'main' into 29048_task_1

Jaimin020 committed:
Linting check fix

Jaimin020 commented:
Hii, @StefanieSenger 

I have made all the changes you mentioned. When I committed to the repo, there was one check failure regarding 'cross_validation'. However, all checks passed when I tried it locally (see the image below for reference).

<img width="1138" alt="Screenshot 2024-06-10 at 9 59 47 PM" src="https://github.com/scikit-learn/scikit-learn/assets/48099245/cfbb6670-a184-4b4b-98fb-0b76633da7be">

Jaimin020 committed:
Minor Bug fix.

The modification is completed to pass all tests.

StefanieSenger reviewed (commented):
Nice, thank you, @Jaimin020. 
I've made a little comment about the param validation of the np.nan value. Apart from that it looks good to me. I'm not a maintainer though. @glemaitre, maybe you want to have a look?

Jaimin020 committed:
Merge branch 'scikit-learn:main' into 29048_task_1

Jaimin020 committed:
Added "np.nan" instead of "nan"

Jaimin020 commented:
Hii, @StefanieSenger and @glemaitre

I've implemented all the changes suggested by @StefanieSenger. @glemaitre, please review this PR and let me know if any further modifications are needed.

Jaimin020 committed:
Merge branch 'scikit-learn:main' into 29048_task_1

Jaimin020 committed:
Updates Done

All modifications done.

Jaimin020 commented on sklearn/metrics/_classification.py:
```diff
@@ -154,10 +154,16 @@ def _check_targets(y_true, y_pred):
         "normalize": ["boolean"],
         "sample_weight": ["array-like", None],
+        "zero_division": [
+            Options(Real, {0, 1, np.nan}),
```
Done

Jaimin020 commented on doc/whats_new/v1.6.rst:
```diff
@@ -138,6 +138,13 @@ Changelog
   :pr:`29067` by :user:`Guillaume Lemaitre <glemaitre>`.
 
+:mod:`sklearn.metrics`
+..............................
```
Done

Jaimin020 commented on sklearn/metrics/_classification.py:
```diff
@@ -218,6 +231,16 @@ def accuracy_score(y_true, y_pred, *, normalize=True, sample_weight=None):
+
     # Compute accuracy for each possible representation
     y_type, y_true, y_pred = _check_targets(y_true, y_pred)
     check_consistent_length(y_true, y_pred, sample_weight)
```
Done

Jaimin020 commented on sklearn/metrics/_classification.py:
```diff
@@ -181,6 +187,13 @@ def accuracy_score(y_true, y_pred, *, normalize=True, sample_weight=None):
+
+        Notes:
+        - If set to "warn", this behaves like a 0.0 input, but a warning is also
+        raised.
```
Done

Jaimin020 commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -215,6 +215,29 @@ def test_classification_report_zero_division_warning(zero_division):
             assert not record
 
 
+def test_accuracy_score_zero_division_warning():
```
I have revised the test case according to your suggestions. Please review it.

Jaimin020 commented:
Hii @glemaitre,

I have implemented all the changes you requested. Please review it.

Jaimin020 committed:
Merge branch 'main' into 29048_task_1

Jaimin020 committed:
Merge branch 'main' into 29048_task_1

Jaimin020 committed:
Merge branch 'main' into 29048_task_1

Jaimin020 committed:
Minor Bug Fix

Jaimin020 committed:
Minor Bug Fix

Jaimin020 committed:
Merge branch 'main' into 29048_task_1

Jaimin020 committed:
Merge branch 'main' into 29048_task_1

Jaimin020 committed:
Merge branch 'main' into 29048_task_1

Jaimin020 committed:
Merge branch 'main' into 29048_task_1

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/Jaimin020/29213

glemaitre committed:
nitpicks

glemaitre reviewed (approved):
LGTM @Jaimin020.

I quickly pushed a commit to merge main in the branch and just modify the message related to skipping the test that is rather a nitpick.

We will need a second approval.

Jaimin020 commented:
> LGTM @Jaimin020.
> 
> I quickly pushed a commit to merge main in the branch and just modify the message related to skipping the test that is rather a nitpick.
> 
> We will need a second approval.

Thanks @glemaitre for the quick update and merge with the main branch.

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into 29048_task_1
